LinkedList.h

```diff
@@ -0,0 +1,83 @@
+#pragma once
+
+#include <pthread.h>
+
+namespace jni 
+{
+	template<class T>
+	class LinkedList 
+	{
+	public:
+		LinkedList()
+		{
+			pthread_mutexattr_t attr;
+			pthread_mutexattr_init(&attr);
+			pthread_mutexattr_settype(&attr, PTHREAD_MUTEX_RECURSIVE);
+			pthread_mutex_init(&lock, &attr);
+			pthread_mutexattr_destroy(&attr);
+			head = NULL;
+		}
+
+		~LinkedList()
+		{
+			pthread_mutex_destroy(&lock);
+		}
+
+		void Add(T* obj)
+		{
+			pthread_mutex_lock(&lock);
+			head = new Link(obj, head);
```

Comment:
Now that you know you only deal with Proxy and Class you can remove this heap allocation and make Class and Proxy linkable. You do this by moving the next ptr into Class and Proxy. That way you can drop Link completely.